### PR TITLE
[mlir][ArmSME] Switch to an attribute-based tile allocation scheme

### PR DESCRIPTION
**Pull request by MacDue**
This reworks the ArmSME dialect to use attributes for tile allocation. This has a number of advantages and corrects some issues with the previous approach:

 * Tile allocation can now be done ASAP (i.e. immediately after `-convert-vector-to-arm-sme`)
 * SSA form for control flow is now supported (e.g.`scf.for` loops that yield tiles)
 * ArmSME ops can be converted to intrinsics very late (i.e. after lowering to control flow)
 * Tests are simplified by removing constants and casts
 * Avoids correctness issues with representing LLVM `immargs` as MLIR values
    - The tile ID on the SME intrinsics is an `immarg` (so is required to be a compile-time constant), `immargs` should be mapped to MLIR attributes (this is already the case for intrinsics in the LLVM dialect)
    - Using MLIR values for `immargs` can lead to invalid LLVM IR being generated (and passes such as -cse making incorrect optimizations)

As part of this patch we bid farewell to the following operations:

```mlir
arm_sme.get_tile_id : i32
arm_sme.cast_tile_to_vector : i32 to vector<[4]x[4]xi32>
arm_sme.cast_vector_to_tile : vector<[4]x[4]xi32> to i32
```

These are now replaced with:
```mlir
// Allocates a new tile with (indeterminate) state:
arm_sme.get_tile : vector<[4]x[4]xi32>
// A placeholder operation for lowering ArmSME ops to intrinsics:
arm_sme.materialize_ssa_tile : vector<[4]x[4]xi32>
```

The new tile allocation works by operations implementing the `ArmSMETileOpInterface`. This interface says that an operation needs to be assigned a tile ID, and may conditionally allocate a new SME tile.

Operations allocate a new tile by implementing...
```c++
std::optional<arm_sme::ArmSMETileType> getAllocatedTileType()
```
...and returning what type of tile the op allocates (ZAB, ZAH, etc).

Operations that don't allocate a tile return `std::nullopt` (which is the default behaviour).

Currently the following ops are defined as allocating:
```mlir
arm_sme.get_tile
arm_sme.zero
arm_sme.tile_load
arm_sme.outerproduct // (if no accumulator is specified)
```

Allocating operations become the roots for the tile allocation pass, which currently just (naively) assigns all transitive uses of a root operation the same tile ID. However, this is enough to handle current use cases.

Once tile IDs have been allocated subsequent rewrites can forward the tile IDs to any newly created operations.

### CODE REVIEW

File: mlir/tools/mlir-query/mlir-query.cpp [hunk 1]
```diff
@@ -21,9 +21,9 @@
 using namespace mlir;
 
 // This is needed because these matchers are defined as overloaded functions.
-using HasOpAttrName = detail::AttrOpMatcher(StringRef);
-using HasOpName = detail::NameOpMatcher(StringRef);
-using IsConstantOp = detail::constant_op_matcher();
+using HasOpAttrName = mlir::detail::AttrOpMatcher(StringRef);
+using HasOpName = mlir::detail::NameOpMatcher(StringRef);
+using IsConstantOp = mlir::detail::constant_op_matcher();
```
Comment (banach-space) [inline mlir/tools/mlir-query/mlir-query.cpp:27]: Unrelated changes?

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td [hunk 2]
```diff
@@ -21,6 +21,99 @@ include "mlir/Interfaces/SideEffectInterfaces.td"
...
+
+    | Tile Vector Types                                                       | Possible Tile IDs   |
+    |-------------------------------------------------------------------------|---------------------|
+    | `vector<[16]x[16]xi8>`                                                  | 0                   |
+    | `vector<[8]x[8]xi16>`, `vector<[8]x[8]xf16>`, or `vector<[8]x[8]xbf16>` | 0 and 1             |
+    | `vector<[4]x[4]xi32>` or `vector<[4]x[4]xf32>`                          | 0 to 3 (inclusive)  |
+    | `vector<[2]x[2]xi64>` or `vector<[2]x[2]xf64>`                          | 0 to 7 (inclusive)  |
+    | `vector<[1]x[1]xi128>`                                                  | 0 to 15 (inclusive) |
+
+    Operations that allocate a new tiles (such as arm_sme.get_tile), are used as
```
Comment (banach-space) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:55]:
```suggestion
    Operations that allocate a new tile (such as arm_sme.get_tile), are used as
```

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td [hunk 3]
```diff
@@ -21,6 +21,99 @@ include "mlir/Interfaces/SideEffectInterfaces.td"
...
+      I32EnumAttrCase<"ZAQ", 4, "za.q">,
+    ]>{
+  let cppNamespace = "mlir::arm_sme";
+  let genSpecializedAttr = 0;
+}
+
+def ArmSMETileOpInterface : OpInterface<"ArmSMETileOpInterface"> {
+  let description = [{
+    An interface for operations that use or allocate Arm SME tiles. These
+    operations need to be assigned a tile ID an i32 attribute, which specifies
```
Comment (banach-space) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:43]:
```suggestion
    operations need to be assigned a tile ID, an i32 attribute, which specifies
```

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td [hunk 4]
```diff
@@ -21,6 +21,99 @@ include "mlir/Interfaces/SideEffectInterfaces.td"
...
+      /*methodBody=*/[{}],
+      /*defaultImpl=*/ [{
+        if (!tileId)
+          return;
+        ::mlir::Operation* op = this->getOperation();
+        op->setAttr("tile_id", tileId);
+      }]
+    >,
+    InterfaceMethod<
+      "Returns the (possibly null) tile ID assigned to this operation.",
```
Comment (banach-space) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:74]: Does `null` correspond to  "tile allocation hasn't been run yet"?

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td [hunk 5]
```diff
@@ -21,6 +21,99 @@ include "mlir/Interfaces/SideEffectInterfaces.td"
...
+      /*methodName=*/"getTileId",
+      /*arguments=*/(ins),
+      /*methodBody=*/[{}],
+      /*defaultImpl=*/ [{
+        ::mlir::Operation* op = this->getOperation();
+        return op->getAttrOfType<mlir::IntegerAttr>("tile_id");
+      }]
+    >,
+    InterfaceMethod<
+      "The type of tile this operation allocates (or none)",
```
Comment (banach-space) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:85]: What does "none" correspond to?

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td [hunk 6]
```diff
@@ -21,6 +21,99 @@ include "mlir/Interfaces/SideEffectInterfaces.td"
...
+    // A helper to create a new operation and propagate this operations tile ID.
+    template<typename T, typename... Args>
+    T createOpAndForwardTileId(::mlir::RewriterBase& rewriter, ::mlir::Location loc, Args &&...args) {
+      auto op = rewriter.create<T>(loc, std::forward<Args>(args)...);
+      if (auto tileOp = ::llvm::dyn_cast<ArmSMETileOpInterface>(op.getOperation()))
+        tileOp.setTileId($_op.getTileId());
+      return op;
+    }
+
+    // A helper to replace this operation and forward any tile ID.
```
Comment (banach-space) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:107]: nice! 
```suggestion
    // A helper to replace this operation and forward its tile ID (if present).
```
?

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td [hunk 7]
```diff
@@ -162,125 +222,67 @@ def ArmSME_CombiningKindAttr : EnumAttr<ArmSME_Dialect, CombiningKind,
...
+def GetTile : ArmSME_Op<"get_tile", [ArmSMETileOpInterface]> {
+  let summary = "Returns a SME virtual tile";
   let description = [{
-    A `cast_tile_to_vector` operation does a cast from a tile id to a 2-d
-    scalable vector type, which represents an SME "virtual tile". This would
-    normally be used when lowering operations that return "virtual tile" vector
-    types to model the output. This is required to preserve dataflow as SME
-    intrinsics have no return values.
+    Allocates a new SME "virtual tile" within a function. The contents of the
+    tile returned from this operation undefined.
```
Comment (banach-space) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:229]: Why "within a function"?

Also:
```suggestion
    Allocates a new SME "virtual tile" within a function. The contents of the
    tile returned from this operation are undefined.
```

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td [hunk 8]
```diff
@@ -162,125 +222,67 @@ def ArmSME_CombiningKindAttr : EnumAttr<ArmSME_Dialect, CombiningKind,
...
+  let summary = "SME tile placeholder";
   let description = [{
-    A `get_tile_id` operation returns a scalar integer representing an SME
-    "virtual tile" id. The bitwidth of the scalar indicates the element
-    bitwidth of the "virtual tile".
-
-    The scope of a tile id is a function and cannot be passed or returned from
-    functions.
+    A placeholder to preserve dataflow while lowering to SME intrinsics (which
+    do not take or return tile values). This operation is intended to be DCE'd
```
Comment (banach-space) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:271]:
```suggestion
    do not take or return SME virtual tile values). This operation is intended to be DCE'd
```

File: mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp [hunk 9]
```diff
@@ -107,73 +109,85 @@ enum class TileMask : unsigned {
...
-                    rewriter.getI32IntegerAttr((unsigned)tilesInUse));
+    func->setAttr(kTilesInUseAttr,
+                  rewriter.getI32IntegerAttr((unsigned)tilesInUse));
+
+    // Find all the ops that (transitively) depend on this tile.
+    SetVector<Operation *> dependantOps;
+    getForwardSlice(tileOp.getOperation(), &dependantOps);
+
+    // Set all operations to use the same tile ID.
+    // This is a navie tile allocation scheme, but works for common cases.
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:183]:
```suggestion
    // This is a naive tile allocation scheme, but works for common cases.
```
Why is it naive? Could you elaborate in the comment? My understanding is that all ops that depend on this tile _have to_ use the same `tile_id`. Is that correct?

File: mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp [hunk 10]
```diff
@@ -107,73 +109,85 @@ enum class TileMask : unsigned {
...
   }
 }
 
-/// Allocates a tile to 'tileID' or returns an error if there are no tiles left.
-static LogicalResult getTile(GetTileID tileIDOp, TileMask &tilesInUse,
-                             unsigned &tileID) {
-  auto masks = getMasks(tileIDOp.getType());
-  for (const auto &it : llvm::enumerate(masks)) {
-    const auto tileMask = it.value();
+/// Allocates a tile to 'tileId' or returns an error if there are no tiles left.
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:139]: [nit] `tileId` feels like an implementation detail that I'd skip in a comment documenting a method. Perhaps "Allocates and returns a tile id. Returns an error if there are no tiles left".

Perhaps also rename - otherwise it clashes with `arm_sme.get_tile`.

**Review by banach-space — COMMENTED** [review]
Looks great, thanks! There's quite a bit of churn, but sadly that's unavoidable. 

I've left a few minor comments - nothing major. I'd like to take another look though - this is quite dense 😅 .

File: mlir/tools/mlir-query/mlir-query.cpp (hunk 1 above)
Comment (MacDue) [reply mlir/tools/mlir-query/mlir-query.cpp:27]: I'm not sure... For some reason these are reported as ambiguous if I don't prefix them with `mlir::` now, not worked out why yet.

File: mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp (hunk 9 above)
Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:183]: I think that's correct. I more mean that to correctly allocate tiles in all cases would require more inspection of the program, and inserting new ops like tile moves. 

For example you could end up with a case like:
```
%tile = scf.if %some_cond -> vector<[4]x[4]xi32> {
   scf.yield %tileA {tile_id = 0} : vector<[4]x[4]xi32>
} else {
   scf.yield %tileB {tile_id = 1} : vector<[4]x[4]xi32>
}
```
Which would need to become (something like):
```
%ifResult = arm_sme.get_tile  {tile_id = 2} : vector<[4]x[4]xi32>
%tile = scf.if %some_cond -> vector<[4]x[4]xi32> {
   arm_sme.move_tile %ifResult <- %tileA : vector<[4]x[4]xi32>
   scf.yield %ifResult {tile_id = 2} : vector<[4]x[4]xi32>
} else {
   arm_sme.move_tile %ifResult <- %tileB : vector<[4]x[4]xi32>
   scf.yield %tileB {tile_id = 2} : vector<[4]x[4]xi32>
}

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td (hunk 7 above)
Comment (MacDue) [reply mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:229]: Tiles are only allocated with respect to a function right now, but they don't necessarily need to be, MLIR is pretty high-level in SCF, so tiles could be allocated in scopes (like in C).. which could be a something that's useful.

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td (hunk 4 above)
Comment (MacDue) [reply mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:74]: Yep, just that no tile ID has been allocated yet.

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td (hunk 5 above)
Comment (MacDue) [reply mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:85]: It means this operation does not allocate a tile.

File: mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/matmul-transpose-a.mlir [hunk 11]
```diff
@@ -3,10 +3,10 @@
 // RUN:   -one-shot-bufferize="bufferize-function-boundaries" -canonicalize \
 // RUN:   -enable-arm-streaming="streaming-mode=streaming-locally za-mode=new-za" \
 // RUN:   -convert-vector-to-arm-sme -convert-arm-sme-to-scf \
-// RUN:   -convert-vector-to-scf -cse -arm-sve-legalize-vector-storage \
+// RUN:   -convert-vector-to-scf -allocate-arm-sme-tiles -cse -arm-sve-legalize-vector-storage \
```
Comment (c-rhodes) [inline mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/matmul-transpose-a.mlir:6]: move before `-convert-vector-to-scf`?

File: mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/fill-2d.mlir [hunk 12]
```diff
@@ -4,9 +4,9 @@
 // RUN:   -lower-vector-mask \
 // RUN:   -one-shot-bufferize="bufferize-function-boundaries" \
 // RUN:   -enable-arm-streaming="streaming-mode=streaming-locally za-mode=new-za" \
-// RUN:   -convert-vector-to-arm-sme -convert-arm-sme-to-scf \
+// RUN:   -convert-vector-to-arm-sme -convert-arm-sme-to-scf -allocate-arm-sme-tiles  \
```
Comment (c-rhodes) [inline mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/fill-2d.mlir:7]: you mentioned `-allocate-arm-sme-tiles` can happen immediately after `-convert-vector-to-arm-sme`, is this the case here?

**Review by c-rhodes — COMMENTED** [review]
Great work Ben, I've left a few minor questions / comments but otherwise LGTM, really polished

File: mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp [hunk 13]
```diff
@@ -506,6 +496,7 @@ struct TileVectorPrintOpConversion : public OpRewritePattern<vector::PrintOp> {
       rewriter.setInsertionPointToStart(forOp.getBody());
       // Extract the current row from the tile.
       Value rowIndex = forOp.getInductionVar();
+      // FIXME: Forward tile IDs.
```
[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp:498]: It's because `vector.print` is not a `ArmSMETileOp`, so unlike the other ops this would require looking at the use-def chain of the tile.

[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp:498]: As it's only for debug, I didn't think it was crucial :smile:

[the remark this answers is not in the document]
Comment (c-rhodes) [reply mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp:498]: ok thanks for the clarification. I thought this was currently used in the integration tests?

[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp:498]: It's used it some, but not all. The simple workaround is to do `-allocate-arm-sme-tiles` after `-convert-arm-sme-to-scf` (since this lowers it to `ArmSMETileOps`).

**Review by c-rhodes — APPROVED** [review]
LGTM cheers

File: mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp:498]: Done :+1:

[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp:498]: Another solution is just to move this to `convert-vector-to-arm-sme`, which I think is fine (as we already generate loops for arith ops in there).

[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp:498]: (That can be a follow up patch though)

[the remark this answers is not in the document]
Comment (c-rhodes) [reply mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp:498]: This got me thinking. The tile allocation works by assigning tile IDs for the root operations that implement the `ArmSMETileOpInterface` and allocate tiles. An operation allocates a tile if it implements `getAllocatedTileType`, this includes:
```mlir
arm_sme.get_tile
arm_sme.zero
arm_sme.tile_load
arm_sme.outerproduct // (if no accumulator is specified)
```
the tile ID is then forwarded to any uses of the root that also implement the `ArmSMETileOpInterface`. Is this constraint necessary? If the tile ID could be forwarded to any use (i.e. non-ArmSME ops) could it fix this problem?

An alternative of course is to mirror ops such as `vector.print` in the ArmSME dialect with ops that implement this interface.

File: mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp [hunk 14]
```diff
@@ -32,24 +32,41 @@ using namespace mlir;
...
+  return tileId;
+}
+
+struct GetTileConversion : public ConvertOpToLLVMPattern<arm_sme::GetTileOp> {
+  using ConvertOpToLLVMPattern<arm_sme::GetTileOp>::ConvertOpToLLVMPattern;
+
+  LogicalResult
+  matchAndRewrite(arm_sme::GetTileOp getTile, OpAdaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    rewriter.replaceOpWithNewOp<arm_sme::MaterializeSSATileOp>(
```
Comment (c-rhodes) [inline mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp:49]: Could `arm_sme.materialize_ssa_tile` be replaced with `llvm.mlir.undef`?

File: mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp:498]: I think the correct fix here is just to more the vector.print lowering into `-convert-vector-to-arm-sme` as it's a vector op lowering, not an ArmSME op lowering.

[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp:498]: Which is how other similar tile ops are lowered (e.g. vector.splat, vector.broadcast, etc). Using an interface allows setting/getting the tile IDs via methods (and validating them), it also avoids possible conflicts with other dialects (and prevents the IR from getting very messy.

File: mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp (hunk 14 above)
Comment (MacDue) [reply mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp:49]: No, `llvm.mlir.undef` only suports LLVM types, and there's no LLVM type for SME tiles.

File: mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (c-rhodes) [reply mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp:498]: > I think the correct fix here is just to more the vector.print lowering into `-convert-vector-to-arm-sme` as it's a vector op lowering, not an ArmSME op lowering (no extra op required).

Ok, I think that's fine for now. As you mentioned, this doesn't have to be part of this patch. We may also have to revisit loop materialization in `-convert-vector-to-arm-sme`. When loads/stores were first implemented loop materialization was done during lowering to intrinsics, `-convert-arm-sme-to-scf` was introduced to break that up such that the intrinsic lowerings were 1-1. It's not quite the same when lowering Vector ops to ArmSME, but there may be value in getting the granularity right, that could necessitate more ops in ArmSME however, so some cost there at least.

File: mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp [hunk 15]
```diff
@@ -107,73 +107,151 @@ enum class TileMask : unsigned {
...
+      // Don't traverse outside a function.
+      if (llvm::isa<FunctionOpInterface>(parent))
+        continue;
+      traverseCorrespondingValues(user->getOperands(), parent->getResults());
+    } else {
+      // Otherwise, assume users of _any_ result are dependant.
+      for (Value result : user->getResults())
+        findDependantOps(result, dependantOps);
+    }
+  }
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:195]: Please avoid nested `if` :)

https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

**Review by banach-space — COMMENTED** [review]
Mostly nits. I'd like to take another look at tile allocation - just to better understand the limitations. The overall design makes a lot of sense.

File: mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp [hunk 16]
```diff
@@ -107,73 +107,151 @@ enum class TileMask : unsigned {
...
       tilesInUse = TileMask::kNone;
 
-    unsigned tileID;
-    if (failed(getTile(tileIDOp, tilesInUse, tileID)))
-      return failure();
+    auto tileId = allocateTileId(*tileType, tilesInUse);
+    if (failed(tileId))
+      return tileOp.emitError("ran out of SME virtual tiles!");
+
+    func->setAttr(kTilesInUseAttr,
```
Comment (joker-eph) [inline mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:221]: Please use consistently `getDiscardableAttr` and `setDiscardableAttr` in newly written code.

You code needs to be resilient to these attribute being "discarded", by the nature of it!

File: mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp (hunk 15 above)
Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:195]: I'm not really sure how that applies here? My understanding of early exits is if you were doing something like:
```
if (A && B && C) {
  // Do thing
}
return null;
```
It's nicer to rewrite it as:
```
if (!A || !B || !C)
  return null;

// Do thing
```
Which make sense :)

But this is checking a number of independent cases, so seems most natural as if/else?

Comment (joker-eph) [reply mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:195]: Right it's not a clear cut here. If you wanted something more structured, there is `TypeSwitch`:

```
TypeSwitch<Operation *>(user)
          .Case<cf::BranchOp>(
              [&](auto brancOp) {
                // (CF) Follow branch.
               traverseCorrespondingValues(branchOp.getDestOperands(),
                                  branchOp.getDest()->getArguments());
           }).Case<...>...
```

File: mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp (hunk 16 above)
Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp]: Done :+1:

File: mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp (hunk 15 above)
Comment (banach-space) [reply mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp]: > I'm not really sure how that applies here? My understanding of early exits is if you were doing something like:

You can also do this:
```cpp
if  (A) {
 // Do thing
 continue;
 }
 
if (B) {
  // Do thing
  continue;
}
```

`TypeSwitch` is obviously much nicer :)

**Review by banach-space — APPROVED** [review]
Makes a lot of sense, LGTM :)

It's been a huge effort - thank you for working on this! This change is a significant improvement the SME support in MLIR 🙏🏻 🙇🏻 .

(please wait for +1 from @joker-eph before landing this)

**Comment by joker-eph** [comment]
> (please wait for +1 from @joker-eph before landing this)

I didn't look deeper than the high-level drive-by comments I made before, so please don't wait for me :)

**Review by dcaballe — COMMENTED** [review]
Awesome! This is a big improvements! Thanks!
I have just one design comment.

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td [hunk 17]
```diff
@@ -162,125 +230,67 @@ def ArmSME_CombiningKindAttr : EnumAttr<ArmSME_Dialect, CombiningKind,
...
-  let summary = "Returns an SME \"virtual tile\" id";
+def MaterializeSSATileOp : ArmSME_Op<"materialize_ssa_tile", [Pure]> {
+  let summary = "SME tile placeholder";
   let description = [{
-    A `get_tile_id` operation returns a scalar integer representing an SME
-    "virtual tile" id. The bitwidth of the scalar indicates the element
-    bitwidth of the "virtual tile".
+    A placeholder to preserve dataflow while lowering to SME intrinsics (which
+    do not take or return SME virtual tile values). This operation is intended
+    to be DCE'd once all ArmSME operations have been lowered.
```
[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:280]: This idea is shown in: https://github.com/MacDue/llvm-project/blob/b9064adf0d6a53c61a907fc2e61f51997a651eab/mlir/test/Dialect/ArmSME/canonicalize.mlir (where the `arm_sme.materialize_ssa_tile` op and the unused tile SSA values fold way).
